### DISCUSSION TITLE
Restart with deformed and pre-stressed body

### DISCUSSION THREAD

**Discussion by feltner515** [idaholab/moose#24760] [Q&A Modules: Solid mechanics] [answered]
Hi Moose forum!

I am having issues with utilizing restart in the context of the tensor mechanics module of moose. Basically, my goal is to begin with a simulation where an elastic indenter is driven into an elasto-plastic substrate. This works no problem. The state of the substrate at the end of simulation 1 should be the initial condition in simulation 2, where the original indenter is deleted and replaced with another indenter at a different position and of a different size. I can project the stress, strain, and displacement information from simulation 1 onto simulation 2 by using auxvariables and variables respectively, but as soon as the simulation begins, the substrate reverts to a stress free state, but remains displaced. I think that I am missing some initialization within either the tensor mechanics master kernel or the plasticity auxkernels, but am having trouble finding where the issue is. Any help would be greatly appreciated!

[TwoDIndenter.zip](https://github.com/idaholab/moose/files/11820059/TwoDIndenter.zip)

Thank you

Langdon

**Comment by hugary1995** [comment, marked as answer]
Hi Langdon, elastoplasticity has history dependent internal variables such as the plastic strain. If you don't initialize the plastic strain in the second simulation properly, the substrate will become elastic and reverts to a (nearly) stress free state.

**Reply by feltner515** [reply]
Hello and thank you for your response! Within the tensor mechanics master kernel, are the elastic and plastic strain stored as separate auxiliary variables that I can make sure are written to the exodus file? I couldn't find in the documentation a convenient way to initialize those as separate entities.

**Reply by GiudGiud** [reply]
Hello

They are not written to the exodus file, which is why you need to use the Checkpoint restart instead of the exodus one
Use checkpoint = true in [Outputs]
Guillaume

**Reply by feltner515** [reply]
Got it, but is there a way I can make the checkpoint work with changing the size and location of the indenter from simulation-to-simulation? My understanding is that the checkpoint restart does not play with changing the mesh. I suppose I can re-break the mesh into blocks and rescale the mesh of the indenter and translate it at each time-step. Do you think this is the easiest course of action?

**Reply by GiudGiud** [reply]
no not currently. Checkpoint cannot be used if the mesh is changed. 

You will have to try to restart in the same mesh, then modify the mesh

**Reply by feltner515** [reply]
Got it, thank you very much for your time!

**Reply by hugary1995** [reply]
Yeah, that's why I didn't mention checkpoint at first... I think you are in the right direction. My suggestion for you is to save all history dependent data as auxvariables, and then restart the simulation (with a different indenter location) and manually reinitialize those data from the auxvariables.

To output properties into exodus, add "outputs = exodus" in every material object. It sounds like you already know how to read those variables back in in the subsequent simulation. The final challenge is to modify the source code to "reinitialize" the data. The method you need to modify is called `initQpStatefulProperties`.

**Reply by feltner515** [reply]
Understood. So is the goal of editing initQpStatefulProperties to make sure that the auxvariables I initialize are properly recognized by the kernel as the initial conditions for the computation?

**Reply by hugary1995** [reply]
ya, you'll want to 
1. couple the appropriate auxvariables to the material object and 
2. initialize the values of the history dependent material property using the values of the auxvariables (which are reloaded from the previous simulation).

There may be quite many properties to initialize this way, now that I think more about it...

Now, let's change the topic entirely. There is indeed a way that checkpoint-restart may work for you. The trick is to modify the mesh at run time (that's actually what @GiudGiud was suggesting -- now I understand). So in theory you should be able to
1. Create a mesh that has the substrate + the first indenter + the second indenter
2. In the first simulation, activate the first indenter and deactivate the second indenter.
3. In the second simulation, start from the latest checkpoint, deactivate the first indenter and activate the second indenter.

The way to "activate/deactivate" an indenter (or in other words, part of the mesh) at run time is using the element subdomain modifiers. There is one object called `CoupledVarThresholdElementSubdomainModifer` that you can play with.

**Reply by hugary1995** [reply]
I am very interested to learn whether this element subdomain modifier works for you -- keep me posted. Once it works, I'd like to add an animation to our homepage gallery if you okay with sharing.

**Reply by feltner515** [reply]
That sounds like the way to go! I will do some reading up on that object and try it out. Thank you for all of your help gentlemen!

**Reply by feltner515** [reply]
Hi @hugary1995 

I wanted to write you a quick follow-up on this. For context, I am a PhD student at Purdue University, working on a project with an industrial consortia of about 15 companies. I am building this model to replicate an industrial process, where the size of indenters and number of indentations are sampled from distributions. This will be packaged as part of a larger flow-sheet model we have developed to track process distributions. In extreme cases, this could go up to 3 to 4 hundred indenters with random impact locations, and the lack of symmetry means I will need to make the model 3-dimensional when I deploy it with our industrial partners. I am worried that with the CoupledVarThresholdElementSubdomainModifer, I could run into issues with memory with so many inactive bodies in the mesh, especially with groups in industry that do not have the same computational resources. It would also be nice for me to be able to use any simulation output as a reference state and vary the size and location of the next indenter. Is memory handled the same way for inactive bodies in the mesh as active bodies?

Following this line of reasoning, I think it may be worth it for me to invest the time in editing initQpStatefulProperties and making exodus restarts work for me. Does that sound reasonable to you?

My understanding is that initQpStatefulProperties lives in the MaterialBase.C file. 

Would this edit be useful to the broader Moose community? 

Thank you for your time!

Langdon

**Reply by hugary1995** [reply]
Langdon,

Your concern regarding the memory is reasonable to me.

> Following this line of reasoning, I think it may be worth it for me to invest the time in editing initQpStatefulProperties and making exodus restarts work for me. Does that sound reasonable to you?

That's doable and definitely not a trivial task (also depending on your constitutive model).

> My understanding is that initQpStatefulProperties lives in the MaterialBase.C file.

That's probably defined there. But you need to override the inherited methods in each and every material model with stateful properties.

> Would this edit be useful to the broader Moose community?

If you take the "modifying initQpStatefulProperties" route, while it is useful, it is hard for me to imagine a way to merge your changes into upstream moose.

A more elegant solution is probably to implement "block-restricted" dataLoad. @lindsayad do you see any possibility of implementing something like that? (and sorry for the long thread to follow)

Either way, it may be better if we have a discussion together with the project PI(s). My email is thu@anl.gov.

Gary

**Reply by lindsayad** [reply]
A block-restricted data load sounds difficult. You would need to know exactly how much data in the checkpoint file to skip over

**Reply by loganharbour** [reply]
The above wouldn't be too unreasonable now that we're moving the ES load into moose.
